Implemented socket logic in current code

DikshantK2004 opened the pull request:
Issue: #20 

### Changes
- added socket server logic into the backend
- fixed the check admin bug in removeUserFromRoom function
- replaced taking roomId to taking room code in req.query to make it easier

DikshantK2004 committed:
Added sockets in room controllers

DikshantK2004 committed:
Merge remote-tracking branch 'upstream/main'

VatsalBhuva11 commented on app.js:
```diff
@@ -25,3 +66,31 @@ app.listen(PORT, () => {
     connDB();
 });
 
+app.get('/room/',async  (req, res) => {
+    console.log('get rooms');
+    const rooms = await prisma.room.findMany();
+    res.json(rooms);
+});
+
+app.get('/user/',async  (req, res) => {
+    console.log('get users');
+    const users = await prisma.user.findMany();
+    res.json(users);
+});
+
+app.delete('/room/delete/:code', async (req, res) => {
+    const { code } = req.params;
+    const room = await prisma.room.delete({
+        where: {
+            code: code
+        }
+    });
+    return res.status(200).send("Room deleted");
+});
+
```
please don't write these controllers here. create their Routers in the necessary files in the `/src/routes` folder.

DikshantK2004 replied:
these are just test routes, I'll remove them

VatsalBhuva11 replied:
no keep them

VatsalBhuva11 replied:
we'll anyway have to have these routes for displaying in the frontend.

VatsalBhuva11 commented on src/controllers/room.controllers.js:
```diff
@@ -26,7 +28,8 @@ async function createRoom(req, res) {
             data: {
-                ...req.body,
+                "roomName" : req.body.roomName,
+                "roomDescription" : req.body.roomDescription,
```
the quotations around `roomName` and `roomDescription` are not needed

criticic commented:
> fixed the check admin bug in removeUserFromRoom function

btw @DikshantK2004 i had fixed it in #26

VatsalBhuva11 commented on src/controllers/room.controllers.js:
```diff
@@ -131,17 +151,30 @@ async function removeUserFromRoom(req, res) {
+        if(!possibleCreator)
+        return response_403(res, "This User who claims to be the admin doesn't exist");
+
+        if(!possibleCreator.isCreator || !room.users.find((user) => user.id == req.user.id)) 
```
although this is logically correct, changing it to 
`if (!(possibleCreator.isCreator && room.users.find((user) => ... )))` would be more readable.

VatsalBhuva11 commented on app.js:
```diff
@@ -4,6 +4,47 @@ import morgan from 'morgan';
+    console.log('socket connected');
+
+
+    socket.on("leaveRoom" , (roomId, userName)=>{
```
just to make sure, you're assuming that the client or the browser socket will emit the "leaveRoom" event when a user wants to leave the room right?

DikshantK2004 replied:
It was another testing event, I will remove it now, io.to('/').adapter.on('leave-room') is now handling this, which is triggered automatically when user leaves a room, and logs a message at server that some user has left some room

VatsalBhuva11 replied:
okay cool
there are multiple other such instances of test events that I came across

VatsalBhuva11 commented on src/controllers/room.controllers.js:
```diff
@@ -49,11 +52,27 @@ async function createRoom(req, res) {
+            if(socket.handshake.query.id ==  req.user.id)
+            {
+                socket.join(code);
+                socket.emit("room joined", code);
```
so the frontend will have to listen for the "room joined" event, and display the room page accordingly?

DikshantK2004 replied:
yes

DikshantK2004 committed:
Merge remote-tracking branch 'upstream/main'

VatsalBhuva11 commented on src/controllers/room.controllers.js:
```diff
@@ -154,25 +187,36 @@ async function removeUserFromRoom(req, res) {
+        {
+            if(socket.handshake.query.id ==  userId)
+            {
+                socket.to(roomCode).emit('user kicked', user.name);
```
change this to 'user removed'. kicked sounds aggressive

VatsalBhuva11 commented on src/controllers/room.controllers.js:
```diff
@@ -237,6 +280,22 @@ async function addUserToRoom(req, res) {
+            {
+                socket.join(roomCode);
+                socket.emit("room joined", roomCode);
+                socket.to(roomCode).emit('new join', user.name);
```
make this "io.to" instead of "socket.to"

DikshantK2004 committed:
Made suggested changes

made routers for some test routes, removed leave-room event in socket, added the test code and other small changes

criticic commented on sockets test/test.md:
```diff
@@ -0,0 +1,6 @@
+- Run admin.html and user.html in VS CODE live server(or some other way)
+- Start the server
+- admin.html is supposed to create the room(both files are same, just that tasks are different)
+- user.html is supposed to join the room
```
we could just use express to serve the html file

VatsalBhuva11 replied:
+1. anyway the html file is not all that important as it's for testing, so no need to code this part although it's a small addition.

VatsalBhuva11 commented on src/routes/v1/userAuth.routes.js:
```diff
@@ -6,4 +6,12 @@ const router = express.Router();
 router.route("/register").post(registerUser);
 router.route("/login").post(loginUser);
 
+// another test route
+router.route("/").get( async (req, res) => {
+    const users = await prisma.user.findMany();
+    return res.status(200).json(users);
+});
+
+
+
```
this controller could go in the user.controller.js file. no worries, ill make this change from my side.

VatsalBhuva11 commented on src/routes/v1/room.routes.js:
```diff
@@ -7,14 +7,39 @@ import {
+    return res.json(rooms);
+})
+
+roomRouter.delete('/delete/:code', async (req, res) => {
+    const { code } = req.params;
+    const room = await prisma.room.delete({
+        where: {
+            code: code
+        }
+    });
+    return res.status(200).send("Room deleted");
+});
```
room deletion would also require the user to be authenticated, but since this is for testing, im ignoring it for now
